### ISSUE TITLE
A Self-Learning Modern Computer Science Curriculum

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4218]
A Self-Learning, Modern Computer Science Curriculum<br>
<br>
http://bit.ly/2nVnGcQ<br>
<br>
About 0 Minutes<br>
<br>
<div><div>
<nav>
<h2>Table of Contents</h2>
</nav>
<div>
<h2>
<span>1</span> Introduction</h2>
<div>
<blockquote>
<p>
"The great idea of constructive type theory is that there is no distinction between programs and proofs… Theorems are types (specifications), and proofs are programs"
(Robert Harper)
</p>
</blockquote>
<p>
This is a collection of modern resources on various undergrad level computer science topics, for someone with an interest in theory. Use <a href="https://en.wikipedia.org/wiki/Library_Genesis">LibGen</a> if you can't buy these books. You don't have to do everything here, just the topics of interest to you.
If for any reason you want a condensed version, watch the <i>Great Theoretical Ideas in Computer Science</i> lecture series to see what interests you in the field. Almost all the material in each resource is self contained so will cover necessary background which if you don't have see the <i>preliminaries</i>. </p>
</div>
<div>
<h3>
<span>1.1</span> Studying Strategies</h3>
<div>
<p>
On Isaac Newton's iteration method to self-learn geometry:
</p>
<blockquote>
<p>
"He bought Descartes' Geometry and read it by himself .. when he was got over 2 or 3 pages he could understand no farther, than he began again and got 3 or 4 pages father till he came to another difficult place, than he began again and advanced farther and continued so doing till he made himself master of the whole without having the least light or instruction from anybody"
(King's Cam.,Keynes MS 130.10,fol. 2/v/) </p>
</blockquote>
<p>
An anecdote on the studying method that works for me: </p>
<ol>
<li>watch a lecture and pay attention to what's going on instead of taking notes. </li>
<li>afterwards take notes from the course slides or skip through the recorded lecture and take notes.
<ul>
<li>if definitions are presented in a lecture try and write my own examples, if a proof is given try to prove it myself. </li>
</ul>
</li>
<li>skim through the assigned reading, writing my own examples for any definitions in the reading and adding those to my notes
<ul>
<li>if exercises are in the reading try them, reading the relevant text chapters if I can't finish an exercise
</li>
</ul>
</li>
<li>proceed to that week's course assignments or labs, or any exams available
</li>
</ol>
<p>
Some readings are so dense and filled with definitions I had to spend a lot of time reading each page carefully to understand it. I also heavily used math or cs stackexchange in the beginning.
Finally, from my notes I'll give a lecture to an invisible class with minimal jargon until I'm satisfied I understand that chapter or lecture. It's called the Feynman technique and is explained <a href="http://calnewport.com/blog/2012/10/26/mastering-linear-algebra-in-10-days-astounding-experiments-in-ultra-learning/">here</a> "If you can’t, out loud or on paper, explain the idea without confusion or contradiction, stop and figure it out right there".
Eventually you will get stuck on a problem set that has errata you forgot to look up. After working at this for an hour or so and being confused what's going on you'll remember to always look up the errata for the assigned reading. </p>
<p>
If you ignore the copious amounts of marketing on his site, Cal Newport has some other interesting anecdotes on studying, such as how he was able to get the <a href="http://calnewport.com/blog/2008/11/25/case-study-how-i-got-the-highest-grade-in-my-discrete-math-class/">best</a> grade in his Discrete Mathematics class, and the rest of the site is full of advice on studying, how to schedule yourself and deliberate <a href="http://calnewport.com/blog/2010/01/06/the-grandmaster-in-the-corner-office-what-the-study-of-chess-experts-teaches-us-about-building-a-remarkable-life/">practice</a>. Further anecdotes exist on the importance of deep work and <a href="https://azeria-labs.com/the-importance-of-deep-work-the-30-hour-method-for-learning-a-new-skill/">deliberate</a> practice.
</p>
</div>
</div>
</div>
<div>
<h2>
<span>2</span> Preliminaries</h2>
<div>
<h3>
<span>2.1</span> Intro to Computing</h3>
<div>
<p>
How command flags/switches work, how to setup networking, basic firewall introduction, how package management works (*nix, Windows &amp; MacOS)
</p>
<ul>
<li>
<a href="http://www.cs.cornell.edu/courses/cs2043/2019sp/schedule.html">(Slides) CS 2043 Unix Tools &amp; Scripting</a>
<ul>
<li>Editors such as emacs/vim are introduced <a href="https://www.cs.cmu.edu/~15131/f17/">here</a> </li>
<li>The 1980s book <a href="https://en.wikipedia.org/wiki/The_Unix_Programming_Environment">The Unix Programming Environment</a> by Kernighan &amp; Pike walks through Awk/Grep/Sed and bash scripting
<ul>
<li>This recent <a href="http://kernighan.com/udw.html">book</a> by Kernighan provides a clear explanation on the basics of modern computer hardware, software and communications networks </li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div>
<h3>
<span>2.2</span> A Rigorous Intro to Programming</h3>
<div>
<p>
The best rigorous introduction that doesn't assume any prior background, or assumes your background was taught incorrectly, is <a href="https://papl.cs.brown.edu/2018/index.html">PAPL</a> <i>Programming and Programming Languages</i>. </p>
<p>
This book is an active learning introduction dynamic programming, memoization, types, graphs, closures, testing, parsing, objects, analysis and solving common recurrences plus numerous topics all relevant to pursuing a CS undergrad. PAPL's <a href="https://www.quora.com/What-does-Professor-Shriram-Krishnamurthi-expect-a-student-to-know-after-he-she-works-through-his-book-Programming-and-Programming-Languages-How-long-until-the-student-is-ready-for-an-interview-at-a-top-software-company">authors</a> combined 3 semesters of Brown into one book. The first 2/3 of the book, have recorded <a href="https://brown.hosted.panopto.com/Panopto/Pages/Sessions/List.aspx#folderID=%222559ad35-3fff-4ef1-a2ac-a951011eb91e%22&amp;maxResults=50">lectures</a> from the Brown university class CSCI 0190 <i>Accelerated Introduction to Computer Science</i> and <a href="http://cs.brown.edu/courses/csci0190/2018/assignments.html">assignments</a>.
The lectures have an interesting teaching style, worth watching if you want to see how data driven, evidence based teaching is done at an ivy league university but they compliment the text, not replace it as programming is best learned by doing. You can run everything in a browser with the <a href="https://www.pyret.org/index.html">Pyret</a> online interpreter, local software install isn't needed. This is the only intro to programming course I've seen where order notation is properly explained. This is also the most rigorous introduction
I've seen, the book and course are designed so you can't easily coast through the material and are forced to think. </p>
<p>
The last 1/3 of the book is used for a class on programming languages, with unique <a href="https://cs.brown.edu/courses/cs173/2018/assignments.html">assignments</a>. Prof. Shriram Krishnamurthi co-authored another introductory book <a href="https://htdp.org/">HtDP</a> <i>How to Design Programs</i> and designs public school curriculum at <a href="https://www.bootstrapworld.org/index.shtml">Bootstrap</a> which has had <a href="https://youtu.be/5c0BvOlR5gs">success</a> teaching this style of material.
The design recipes in HtDP are sometimes referred to in PAPL/lectures, you may want to try some of these but completing HtDP is not a prereq or hard dependency for PAPL. </p>
</div>
</div>
<div>
<h3>
<span>2.3</span> Assumed Mathematical Background</h3>
<div>
<blockquote>
<p>
"So I went to Case, and the Dean of Case says to us, it's a all men's school, "Men, look at, look to the person on your left, and the person on your right. One of you isn't going to be here next year; one of you is going to fail." So I get to Case, and again I'm studying all the time, working really hard on my classes, and so for that I had to be kind of a machine.
In high school, our math program wasn't much, and I had never heard of calculus until I got to college. But the calculus book that we had was (in college) was great, and in the back of the book there were supplementary problems that weren't assigned by the teacher. So this was a famous calculus text by a man named George Thomas (<a href="https://www.amazon.com/gp/customer-reviews/R1HJTYVOE94R83/ref=cm_cr_dp_d_rvw_ttl?ie=UTF8&amp;ASIN=0201077795">second edition</a>), and I mention it especially because it was one of the first books published by Addison-Wesley, and I loved this calculus book so much that later I chose Addison-Wesley to be the publisher of my own book.
Our teacher would assign, say, the even numbered problems, or something like that (from the book). I would also do the odd numbered problems. In the back of Thomas's book he had supplementary problems, the teacher didn't assign the supplementary problems; I worked the supplementary problems. I was scared I wouldn't learn calculus, so I worked hard on it, and it turned out that of course it took me longer to solve all these problems than the kids who were only working on what was assigned, at first. But after a year, I could do all of those problems in the same time as my classmates were doing the assigned problems, and after that <i>I could just coast in mathematics, because I'd learned how to solve problems</i>"
(<a href="https://github.com/kragen/knuth-interview-2006">Don Knuth</a> )
</p>
</blockquote>
<p>
See this <a href="https://www.youtube.com/user/njwildberger/playlists">playlist</a> of Math Foundations by UNSW Prof. Wildberger, specifically playlist <a href="https://www.youtube.com/playlist?list=PL5A714C94D40392AB">A</a>, and this lecture series on mathematics <a href="https://www.youtube.com/playlist?list=PL34B589BE3014EAEB">history</a>, which is a walk through of mathematics up to the present. For example, his <a href="https://www.youtube.com/watch?v=K4eAyn-oK4M">lecture</a> on the notation of Limits and epsilon, which you will see in numerous future courses. A great set of short texts, that pair with these lectures and focus on problem solving are the I.M. Gelfand correspondence school texts:
</p>
<ul>
<li>Algebra (<a href="https://www.springer.com/gp/book/9780817636777">Birkhäuser</a>)
</li>
<li>The Method of Coordinates (<a href="https://store.doverpublications.com/0486425657.html">Dover</a>)
</li>
<li>Functions and Graphs (<a href="https://store.doverpublications.com/0486425649.html">Dover</a>)
</li>
<li>Trigonometry (<a href="https://www.springer.com/gp/book/9780817639143">Birkhäuser</a>)
</li>
<li>Sequences, Combinations, Limits (<a href="https://store.doverpublications.com/0486425665.html">Dover</a>)
</li>
</ul>
<p>
There is also a trigonometry <a href="http://www.math.brown.edu/~dkatz/trigbootcamp/">boot camp</a> from Brown. For additional practice there is <a href="https://en.wikipedia.org/wiki/Po-Shen_Loh">Po-Shen Loh's</a> site <a href="https://www.expii.com/">Expii</a>, and Cambridge also offers advice for incoming undergrads on readable <a href="http://www.maths.cam.ac.uk/documents/reading-list.pdf/">texts</a>. The only thing to know about math is it's a learn by doing subject, and you could also just use any typical first year Calculus undergrad text to fix your deficiences in math by going through the exercises and directly looking up whatever you've forgot on Wikipedia or in Sheldon Axler's <i>Precalculus</i>. MIT's <a href="https://ocw.mit.edu/courses/mathematics/18-01sc-single-variable-calculus-fall-2010/">18.01</a> is probably the best choice, you get full lectures, recitations and practice sets.
</p>
</div>
</div>
</div>
<div>
<h2>
<span>3</span> Functional Programming</h2>
<div>
<h3>
<span>3.2</span> Functional Programming &amp; Data Structures</h3>
<div>
<p>
This Cornell OCaml course is totally self contained with it's own <a href="http://www.cs.cornell.edu/courses/cs3110/2018fa/textbook/">textbook</a>. You will want to archive the entire course locally using <a href="https://www.guyrutenberg.com/2014/05/02/make-offline-mirror-of-a-site-using-wget/">wget</a>.
The release code for the assignments wget them <a href="https://www.cs.cornell.edu/courses/cs3110/2018sp/a5/a5.zip">here</a> and change URL to /a4/a4.zip, /a3/a3.zip, they range a0.zip - a5.zip
</p>
<ul>
<li>
<a href="https://www.cs.cornell.edu/courses/cs3110/2018sp/">(Full Course) CS 3110 Data Structures and Functional Programming</a>
<ul>
<li>The free textbook is essentially the lectures, the <a href="https://www.cs.cornell.edu/courses/cs3110/2018sp/">notes</a> have recommended chapters for additional books like <i>Real World Ocaml</i> and have exercises.
</li>
<li>Introduction to Coq, you can extract OCaml from Coq, F* or from <a href="http://www.cs.ru.nl/~gondelman/toolchain.pdf">Why3</a> </li>
<li>Previous semesters have <a href="https://www.cs.cornell.edu/courses/cs3110/2009fa/lecturenotes.asp">notes</a> on asymptotic complexity of OCaml programs and solving recurrences
</li>
<li>Assumes you have already done PAPL or another rigorous introductory CS course
<ul>
<li>See this <a href="https://www.youtube.com/watch?v=-J8YyfrSwTk">talk</a> Effective ML how to properly write interfaces, error handling
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div>
<h2>
<span>4</span> Algebra</h2>
<p>
It is possible in a functional language like ML to do algebra with types, proving two types are isomorphic with the desired properties of reflexivity, symmetry, and transitivity. It's also possible to abstract Lists and Trees into polynomials, as every polynomial looks like a sum of terms.
As you will learn in 15-150 <i>Principles of Functional Programming</i> "most functional datastructures have constant time access near the outer layer of their structure, ie: the head of a list or the root of a tree. However, access at some random point inside the structure is typically linear since looking at some element of a list is
linear in the length of the list, and looking at some element of a tree is linear in the depth of the tree. Datatype derivatives allow constant time access to the entire structure." </p>
<div>
<h3>
<span>4.1</span> Linear Algebra</h3>
<div>
<p>
Wildberger has a series of intuitive <a href="https://www.youtube.com/watch?v=yAb12PWrhV0&amp;list=PLIljB45xT85BhzJ-oWNug1YtUjfWp1qAp">lectures</a> to understand this topic geometrically.
</p>
<ul>
<li>
<a href="http://cs.brown.edu/courses/cs053/current/lectures.htm">(Full Course) CS053 - Coding the Matrix</a>
<ul>
<li>Covers interesting applications like what was shown in <i>Great Theoretical Ideas in Computer Science</i> when you learned how a parity bit works.
</li>
<li>No formal prerequisites except assumes you know how to do basic proofs. Uses Python but you can do this course in <a href="http://ocaml.xyz/chapter/matrix.html">OCaml</a> too.
<ul>
<li>This free <a href="https://www.math.ucdavis.edu/~anne/linear_algebra/index.html">book</a> <i>Linear Algebra - As an Introduction to Abstract Mathematics</i> from UC Davis has plenty of exercises </li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div>
<h3>
<span>4.2</span> Abstract Algebra</h3>
<div>
<p>
Some <a href="https://youtu.be/-Z9XL4B0ROk">videos</a> in this playlist introduce Abstract Algebra, walking through it's foundations distinguishing between descriptions, definitions and specifications of abstract algebraic objects.
</p>
<ul>
<li>
<a href="https://www.math.upenn.edu/~ted/371F14/hw-371SchedTab.html">(Full Course) Math-371 Abstract Algebra</a>
<ul>
<li>Has recorded lectures which are essential as you can quickly get lost in the texts at this level of abstraction
</li>
<li>Uses readings from <a href="https://www.math.upenn.edu/~ted/371F14/info.html#texts">three texts</a> <ul>
<li>There is also this <a href="http://www.extension.harvard.edu/open-learning-initiative/abstract-algebra">Harvard course</a> with excellent recorded lectures and uses readings from <i>Algebra - M. Artin</i> (blue book) as a great compliment to Math-371 lectures </li>
</ul>
</li>
<li>The <a href="http://www.math.hawaii.edu/~lee/algebra/history.html">history</a> of Abstract Algebra by Prof Lee Lady </li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div>
<h2>
<span>5</span> Discrete Mathematics</h2>
<p>
Resources to learn undergrad Discrete Mathematics. </p>
<div>
<h3>
<span>5.1</span> Introduction to Pure Mathematics</h3>
<div>
<p>
Good advice on writing mathematics, introduces LaTeX. Cornell has a similar class with <a href="https://courses.cs.cornell.edu/cs2800/wiki/index.php/CS_2800_Spring_2019">complete</a> lectures notes. </p>
<ul>
<li>
<a href="https://infinitedescent.xyz/">(Book) An Infinite Descent into Pure Mathematics</a>
<ul>
<li>Collection of excellent lecture notes used in CMU's 15-151/21-127 Concepts of Mathematics class </li>
<li>Curriculum based on findings in evidenced based teaching
</li>
<li>List of <a href="https://math.vanderbilt.edu/schectex/commerrs/">Common Errors in Undergrad Math</a> like undistributed cancellations, sign errors, confusion about notation etc.
<ul>
<li>For an introduction to math from a programmer's perspective try <a href="https://pimbook.org/">pimbook</a> instead
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div>
<h3>
<span>5.2</span> Discrete Math with Standard ML</h3>
<div>
<p>
This book covers most of the topics in courses 15-251 and 15-150, and teaches you to write proofs as programs. If you don't know how to program this book will teach you by modeling math functions into SML functions, polynomials into lists, sets into types, creating relations and proving their results with programs. If you really want to understand set theory for combinatorics you should probably read this. </p>
<ul>
<li>
<a href="https://cs.wheaton.edu/~tvandrun/dmfp/">(Book/Lectures) Discrete Mathematics and Functional Programming - Thomas VanDrunen</a>
<ul>
<li>Lectures exist on the author's homepage, this book is used for a one semester university course with additional elective chapters in Graph Theory, Complexity Theory, Automata, etc.
<ul>
<li>The <a href="https://archive.org/details/B-001-002-135">book</a> <i>Concrete Mathematics</i> is an alternative for exercises, as it has answers included for self-study. </li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div>
<h3>
<span>5.3</span> Great Theoretical Ideas in Computer Science</h3>
<div>
<p>
This is a crash course in multiple topics such as Probability, Linear Algebra, Modular Arithmetic, Polynomials, Cryptography and Complexity Theory. 15-251 assumes students have completed <i>An Infinite Descent into Pure Mathematics</i> though it is self contained. </p>
<ul>
<li>
<a href="http://www.cs.cmu.edu/~arielpro/15251f15/schedule.html">(Full Course) 15-251 Great Theoretical Ideas in Computer Science</a>
<ul>
<li>Recorded lectures if you click each topic, backup YouTube <a href="https://www.youtube.com/playlist?list=PLm3J0oaFux3aafQm568blS9blxtA_EWQv">lectures</a> and <a href="http://www.cs.cmu.edu/~15251/notes.html">course notes</a>
</li>
<li>Some more material on the most <a href="http://www.cs.cmu.edu/~15251/schedule.html">recent</a> course schedule such as <a href="http://www.cs.cmu.edu/~15251/notes/induction-pitfalls.pdf">induction pitfalls</a>
<ul>
<li>(Optional) MIT 6.042j <i>Mathematics for Computer Science</i> recorded <a href="https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-042j-mathematics-for-computer-science-fall-2010/video-lectures/">lectures</a> and <a href="https://courses.csail.mit.edu/6.042/spring17/mcs.pdf">course notes</a> </li>
<li>Additional lecture <a href="http://www.cs.cmu.edu/~odonnell/papers/probability-and-computing-lecture-notes.pdf">notes</a> on discrete probability and <a href="https://www.youtube.com/playlist?list=PLIljB45xT85AMigTyprOuf__daeklnLse">lectures</a> about basic probability.
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div>
<h2>
<span>6</span> Computer Systems &amp; Architecture</h2>
<div>
<blockquote>
<p>
"[Computer science] is not really about computers - and it's not about computers in the same sense that physics is not really about particle accelerators, and biology is not about microscopes and Petri dishes…and geometry isn't really about using surveying instruments. Now the reason that we think computer science is about computers is pretty much the same reason that the Egyptians thought geometry was about surveying instruments: when some field is just getting started and you don't really understand it very well, it's very easy to confuse the essence of what you're doing with the tools that you use." (Hal Abelson) </p>
</blockquote>
</div>
<div>
<h3>
<span>6.1</span> x86 Architecture</h3>
<div>
<p>
This covers architecture from a programmer's perspective, such as how to write cache friendly code, assembly, how compilers work, return oriented programming (ROP) to bypass stack protections, the memory hierarchy, and networks. You could read K&amp;R's <i>The C Programming Language</i> for a brief intro, though this course will explain C as you go anyway and fully covers pointers at the assembly language level making it self contained. </p>
<ul>
<li>
<a href="https://www.cs.cmu.edu/~213/schedule.html">(Full Course) 15-213 Intro to Computer Systems (CMU)</a> <ul>
<li>Recorded lectures <a href="http://scs.hosted.panopto.com/Panopto/Podcast/Podcast.ashx?courseid=b96d90ae-9871-4fae-91e2-b1627b43e25e&amp;type=mp4">here</a>
</li>
<li>Get the CS:APP book, 3rd version for x86-64 (beware of the global version <a href="http://csapp.cs.cmu.edu/3e/errata.html">errata</a> )
</li>
<li>The labs (Data Lab, Bomb Lab ect) are on the <a href="http://csapp.cs.cmu.edu/3e/students.html">book website</a> </li>
<li>You can also do systems programming in <a href="http://mlton.org/References.attachments/Shipman02.pdf">Standard ML(pdf)</a> or <a href="https://ocaml.github.io/ocamlunix/index.html">OCaml</a>
<ul>
<li>Try the embedded security CTF <a href="https://microcorruption.com/login">here</a> after you finish the Attack Lab
</li>
<li>(Optional) A <a href="https://os.phil-opp.com/">series</a> of posts on writing an OS in Rust lang. <a href="https://www.red-bean.com/jimb/">Programming Rust</a> is also an excellent systems book </li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div>
<h3>
<span>6.2</span> x86 Operating Systems</h3>
<div>
<p>
MIT teaches OS engineering using a rewrite of the sixth edition Unix(v6) similar to the classic <a href="https://en.wikipedia.org/wiki/Lions%27_Commentary_on_UNIX_6th_Edition,_with_Source_Code">Lions' Commentary</a> but in ANSI-C called <a href="https://pdos.csail.mit.edu/6.828/2018/xv6.html">xv6</a>. The idea is you read the source as you read booklet, to understand the system. There is also a schedule of lecture <a href="https://pdos.csail.mit.edu/6.828/2018/schedule.html">notes</a>.
</p>
<p>
The latest xv6 source and text are available via:
</p>
<ul>
<li>git clone git://github.com/mit-pdos/xv6-public.git
</li>
<li>git clone git://github.com/mit-pdos/xv6-book.git
</li>
</ul>
</div>
</div>
<div>
<h3>
<span>6.3</span> RISC-V Architecture</h3>
<div>
<p>
MIT's 6.004 course covers RISC-V: The free and open RISC Instruction Set Architecture. Take this if you are interested in programming FPGAs or <a href="https://www.seeedstudio.com/sipeed">cheap</a> embedded modules.
</p>
<ul>
<li>
<a href="https://6004.mit.edu/web/spring19/resources/lectures">(Full Course) 6.004 Computation Structures</a>
<ul>
<li>Recorded YouTube lectures <a href="https://www.youtube.com/channel/UC1DcxXg6GkAcp2zk2w7U6qQ/videos">here</a> </li>
<li>Book used is Computer Organization and Design: <a href="https://www.elsevier.com/books/computer-organization-and-design-risc-v-edition/patterson/978-0-12-812275-4">RISC-V Edition</a>
<ul>
<li>A <a href="https://www.morganclaypool.com/toc/cac/1/1">series</a> of papers in Computer Architecture curated by Princeton U (use <a href="https://en.wikipedia.org/wiki/Sci-Hub">sci-hub</a> to access them)
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div>
<h2>
<span>7</span> Networks</h2>
<div>
<p>
How the computing, economic and sociological worlds are connected.
</p>
<ul>
<li>
<a href="https://www.cs.cornell.edu/home/kleinber/networks-book/">(Book) Networks, Crowds and Markets - Kleinberg &amp; Easley</a>
<ul>
<li>Free draft covers introductory Game Theory, matchings, information theory, economics etc.
</li>
<li>A playlist of short lectures from the book authors <a href="https://www.youtube.com/playlist?list=PLPy2pozanw24LBksjW2_vZe5V3dical4K">here</a>
<ul>
<li>Additional advanced material in this preprint 2019 <a href="http://www.cs.cornell.edu/~rafael/networks.pdf">draft</a>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div>
<h3>
<span>7.1</span> Computer Networks</h3>
<div>
<p>
The 15-213 <i>course</i> covers an introduction to computer networks and socket programming. This book explains the principles underlying network architecture design.
</p>
<ul>
<li>
<a href="https://book.systemsapproach.org/">(Book) Computer Networks: A Systems Approach - Larry Peterson and Bruce Davie</a>
<ul>
<li>Creative Commons License, you can build the book yourself from their github repository
<ul>
<li>Additional <a href="http://www.cs.cmu.edu/~15744/syllabus.html">slides</a> and recommended papers to understand the current state of network protocols, methods to test for censorship, topology, etc.
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div>
<h2>
<span>10</span> Practical Data Science</h2>
<div>
<ul>
<li>
<a href="http://www.datasciencecourse.org/lectures/">(Full Course) 15-388 Practical Data Science</a>
<ul>
<li>Introduction to the 'full stack' of data science analysis: data collection and processing, data visualization and presentation, statistical model building using machine learning, and big data techniques for scaling these methods.
</li>
<li>Cornell also has a free <a href="http://www.cs.cornell.edu/courses/cs1380/2018sp/textbook/">text</a> with interactive jupyter notebooks (in Python)
</li>
<li>Jupyter notebook is a <a href="http://www.literateprogramming.com/index.html">literate programming</a> paradigm tool, you can freely write using LaTeX or markdown and run the code cells in any order to not interrupt the flow of documentation. <ul>
<li>This course teaches the Python libraries, but you can do the assignments in <a href="http://ucidatascienceinitiative.github.io/IntroToJulia/">Julia</a> if you want, or experiment in <a href="http://ocaml.xyz/">OCaml</a> </li>
<li>A draft <a href="http://www.cs.cornell.edu/jeh/book11April2014.pdf">book</a> <i>Foundations of Data Science</i> </li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div>
<h2>
<span>11</span> Meta-Linguistic Abstraction</h2>
<div>
<p>
A classic introductory computer science book on thinking about the big picture of programs with abstraction: finding general patterns from specific problems and building programs based on these patterns. An applied example of this is the package manager <a href="http://dustycloud.org/misc/talks/guix/chicagolug_2015/guix_talk.html">GNU Guix</a> and distro GuixSD, which is a GNU implementation of the NixOS <a href="https://arxiv.org/pdf/1305.4584.pdf">functional software deployment model</a>. Package builds, including entire system builds, are declared in one text file. The resulting software deployment is functional: build inputs go in such as compilers, customizations, environments ect, and a reproducible, immutable build comes out with a hashed identifier meaning you can do roll backs to previous successful builds. A recent GUixSD feature <a href="https://arxiv.org/pdf/1709.00833.pdf">gexp</a> (g-expressions) is a good example what can be achieved through syntactic extensions of the Scheme language. You are more likely to benefit from this book after having some programming experience, but no matter what <a href="https://twitter.com/id_aa_carmack/status/350028210551013376?lang=en">level</a> of programmer you are you'll still benefit from SICP. </p>
<ul>
<li>
<a href="http://web.mit.edu/alexmv/6.037/">(Full Course) 6.037 Structure and Interpretation of Computer Programs</a>
<ul>
<li>Recorded lectures from the first edition of SICP <a href="https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-001-structure-and-interpretation-of-computer-programs-spring-2005/video-lectures/">here</a>
</li>
<li>Read the excellent <a href="https://mitpress.mit.edu/sicp/full-text/book/book-Z-H-5.html#%25_chap_Temp_2">forward</a> from SICP on what programming is <ul>
<li>There is a <a href="http://web.mit.edu/~axch/www/test_manager.html">unit test library</a> for SICP as well as an online <a href="http://icampustutor.csail.mit.edu/6.001-public/">tutorial</a>
</li>
<li>SICP in <a href="http://www.neilvandyke.org/sicp-texi/">TeXinfo</a> so you can read it in emacs </li>
<li>Optional read: <a href="http://web.mit.edu/~axch/www/art.pdf">The Art of the Propagator</a> for further adventures in symbolic programming
</li>
<li>If you like Lisp, see Norvig's old AI <a href="https://github.com/norvig/paip-lisp">book</a> for a really great programming craft text
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div>
<h2>
<span>14</span> Isolating Software Failure, Proving Safety and Testing</h2>
<div>
<p>
How to verify software, and strategies of programming that minimize catastrophe during failure. The <a href="https://mitpress.mit.edu/books/little-prover">Little Prover</a> is a good introduction in determining facts about computer programs.
Prof Rob Simmons can be <a href="https://calculem.us/classes/">hired</a> as a tutor to teach you Coq, the curriculum follows the Software Foundations book series below.
</p>
<ul>
<li>
<a href="https://15316-cmu.github.io/index.html">(Lecture Notes) 15-316 Software Foundations of Security and Privacy</a>
<ul>
<li>Covers side channel attacks, provable privacy, web security, sandboxing, assignments are extending an http network server (in C), and spotting bugs. </li>
<li>Read about tests, strategies to safe program design (in OCaml), and proving safety from the 2017 version: git clone <a href="https://github.com/jeanqasaur/cmu-15316-spring17.git">https://github.com/jeanqasaur/cmu-15316-spring17.git</a>
</li>
</ul>
</li>
</ul>
<ul>
<li>
<a href="http://www.cs.princeton.edu/~appel/vfa/">(Book) Verified Functional Algorithms</a> <ul>
<li>Some recorded lectures by Andrew Appel <a href="https://deepspec.org/event/dsss17/lecture_appel.html">here</a>
</li>
<li>Part 3 of the <a href="https://deepspec.org/page/SF/">Deep Specifications</a> interactive book series by Andrew Appel, learn by doing </li>
<li>Assumes you have read <a href="https://softwarefoundations.cis.upenn.edu/current/index.html">these</a> chapters of Software Foundations Part I: Preface, Basics, Induction, Lists, Poly, Tactics, Logic, IndProp, Maps, (ProofObjects), (IndPrinciples)
<ul>
<li>A good introduction to Dependent Types by Dan Licata is <a href="https://youtu.be/LXvP1A97oAM">here</a> or The Little Typer <a href="https://mitpress.mit.edu/books/little-typer">book</a>.
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div>
<h3>
<span>14.1</span> Physical Systems Software Security</h3>
<div>
<ul>
<li>
<a href="http://symbolaris.com/course/fcps16.html">(Full Course) 15-424 Foundations of Cyber-Physical Systems</a>
<ul>
<li>Course (with recorded lectures) if you're interested in programming drones/space shuttles/robots/cars and other things that cannot fail from avoidable errors.
</li>
<li>Self contained, will teach you differential eq but assumes you already have some calculus background. </li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div>
<h2>
<span>15</span> Algorithms</h2>
<div>
<h3>
<span>15.1</span> Introduction to Parallel and Sequential Algorithms</h3>
<div>
<p>
Unlike a traditional introduction to algorithms and data structures, this puts an emphasis on thinking about how algorithms can do multiple things at once instead of one at a time and ways to design to exploit these properties. It uses a language based analysis model to estimate complexity for abstractions such as garbage collection.
</p>
<ul>
<li>
<a href="http://www.parallel-algorithms-book.com/">(Book) Parallel and Sequential Algorithms</a>
<ul>
<li>Complete, self-contained book with exercises used for <a href="https://www.cs.cmu.edu/~./15210/schedule.html">15-210</a>
</li>
<li>Old exams, homework and labs w/solutions are <a href="https://github.com/functionalCS/15210labs">here</a> and done in Standard ML </li>
</ul>
</li>
</ul>
</div>
</div>
<div>
<h3>
<span>15.2</span> Functional Data Structures</h3>
<div>
<p>
"A stroll through intermediate data structures and their associated algorithms, from the point of view of functional programming." Note how pattern matching makes implementing these seemingly complicated structures a much easier task. </p>
<ul>
<li>
<a href="https://cs.uwaterloo.ca/~plragde/flaneries/FDS/">(Book) Functional Data Structures</a> <ul>
<li>Exercises and examples done in OCaml, no formal background beyond a typical introductory CS course
<ul>
<li>A more advanced book <a href="http://www.cambridge.org/ca/academic/subjects/computer-science/programming-languages-and-applied-logic/purely-functional-data-structures?format=PB&amp;isbn=9780521663502">Purely Functional Data Structures</a> by Chris Okasaki uses an imaginary version of SML w/Appendix in Haskell.
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div>
<h3>
<span>15.3</span> Algorithm Design</h3>
<div>
<p>
The typical text for a senior undergrad or graduate course on algorithm core theory and analysis is <i>The Design and Analysis of Algorithms</i> by Kozen (1992). This is an excellent book to study with difficult problems well presented and clearly analyzed by Kozen. If you want a book to practice design, <i>Algorithm Design</i> by Jon Kleinberg and Éva Tardos contains 200+ problems and how to design solutions for them. </p>
<ul>
<li>
<a href="http://people.seas.harvard.edu/~cs224/fall14/lec.html">(Full Course) CS224 Advanced Algorithms</a>
<ul>
<li>Recorded lectures by Prof. Jelani Nelson walking through the design and analysis of online algorithms, linear programming, approximation algorithms, hashing, randomized algorithms.
</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div>
<h2>
<span>16</span> Complexity Theory</h2>
<div>
<h3>
<span>16.3</span> Useful Math for Theoretical CS</h3>
<p>
These scribed lecture <a href="http://www.cs.cmu.edu/~odonnell/toolkit13/">notes</a> give you a diverse background in math useful for theoretical CS, from the excellent book <a href="http://www.nature-of-computation.org/">The Nature of Computation</a> such as these <a href="http://www.cs.cmu.edu/~15751/2016-lecture1.pdf">slides</a> from <i>A Theorist's ToolKit</i> which describe how to find research, how to write math in LaTeX, how to give a talk, resources on math.overflow etc.
</p>
</div>
</div>
<div>
<h2>
<span>17</span> Introduction to Quantum Computing</h2>
<div>
<p>
This is a graduate introduction to quantum computation/information theory, from the perspective of theoretical computer science. Topcoder has quantum algorithm <a href="https://www.topcoder.com/lp/digitalannealer">bounty's</a>.
</p>
<ul>
<li>
<a href="http://www.cs.cmu.edu/~odonnell/quantum18/">(Full Course) 15-859BB: Quantum Computation</a>
<ul>
<li>Recorded lectures on <a href="https://www.youtube.com/playlist?list=PLm3J0oaFux3YL5qLskC6xQ24JpMwOAeJz">YouTube</a>.
</li>
<li>The prereqs are an undergrad background in complexity theory (15-251 or 15-455), linear algebra, and discrete <a href="http://www.cs.cmu.edu/~odonnell/papers/probability-and-computing-lecture-notes.pdf">probability</a>.
</li>
<li>"90% of the understanding of the quantum circuit model is achieved by reviewing three purely 'classical' topics: classical Boolean circuits; reversible classical circuits; and randomized computation"
<ul>
<li>A <a href="https://www.morganclaypool.com/toc/qmc/1/1">series</a> of curated papers on Quantum Computing
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div>
<h3>
<span>17.1</span> Functional Differential Geometry</h3>
<p>
If this topic interests you, Gerald Jay Sussman &amp; Jack Wisdom's <a href="https://mitpress.mit.edu/books/functional-differential-geometry">book</a> teaches differential geometry using Scheme as a foundation for quantum field theory or general relativity. NJ Wildberger additionally has a set of introductory <a href="https://youtu.be/_mvjOoTieTk">lectures</a>.
</p>
</div>
</div>
<div>
<h2>
<span>18</span> Jobs</h2>
<div>
<p>
This <a href="https://triplebyte.com/">service</a> matches your skills to people who want to pay you. Jane Street Capital is a finance tech company that hires functional programmers worldwide, you may want to <a href="https://blogs.janestreet.com/interviewing-at-jane-street/">apply</a> there. There are numerous opportunities to apprentice as a <a href="https://www.nccgroup.trust/us/about-us/careers/current-vacancies/security-consultant/">security researcher</a> or get paid to prove business logic programs like 'smart contracts'. For web programming there exists free programming <a href="https://books.goalkicker.com/">notes</a> for software like NodeJS, which you can write software for in OCaml by compiling it into JS with <a href="https://bucklescript.github.io/">bucklescript</a> or <a href="https://ocsigen.org/js_of_ocaml/3.1.0/manual/overview">js<sub>of</sub><sub>ocaml</sub></a> or just use <a href="https://www.256kilobytes.com/content/show/4399/get-these-dependencies-off-my-lawn-5-tasks-you-didnt-know-could-be-done-with-pure-html-and-css">pure html</a>. There are also bounties for writing features available on topcoder, gitcoin or bountysource or consulting outfits such as toptal or clevertech, which tend to use <a href="https://app.codility.com/programmers/">codility tests</a> for filtering candidates. Do your own research on the current state of these outfits, through glassdoor or other employee review sites. A large collection of remote job listing and consulting platforms is <a href="https://docs.google.com/spreadsheets/d/1JfNAbUX_lN9K3MCNHO15GJtJ5qpk7H9Cl3xTBwv2FR8/htmlview">here</a>.
</p>
<p>
If you enjoy research look up research programming positions at a university near you. You get hired to a grant funded team of grad students and post-doc researchers. Sometimes they are just hiring for people to build
a proper <a href="http://www.cs.cmu.edu/~bam/uicourse/05863fall18/schedule.html">user interface</a> for the doctors, or you could be doing data visualizations, or engineering experimental software. Depending on the work you are doing, they may even write you a letter of recommendation meaning you can use this to apply to a grad school. </p>
</div>
</div>
<div>
<h2>
<span>19</span> Graduate Research in Type Theory</h2>
<p>
Read these <a href="http://www.cs.cmu.edu/~15751/2016-lecture1.pdf">slides</a> from <i>A Theorist's ToolKit</i> on how to find research, how to write math in LaTeX, how to give a talk, where to ask on stackexchange etc.
</p>
</div>
<div>
<h2>
<span>20</span> Graduate Research in Machine Learning/AI</h2>
<p>
Read these <a href="http://www.cs.cmu.edu/~15751/2016-lecture1.pdf">slides</a> from <i>A Theorist's ToolKit</i> on how to find research, how to write math in LaTeX, how to give a talk, where to ask on stackexchange etc. </p>
<div>
<h3>
<span>20.1</span> Graduate Introduction to General AI</h3>
<div>
<p>
These notes <a href="http://www.cs.cmu.edu/~./15381/#schedule">here</a> cover more topics, such as Computer Vision/NLP/AGI and follows <i>AI: A Modern Approach</i> book by Norvig. The 4th edition draft of pseudocode algorithms for the book is <a href="https://github.com/aimacode/aima-pseudocode">here</a>.
</p>
<ul>
<li>
<a href="http://www.cs.cmu.edu/afs/cs/Web/People/15780/#schedule">(Full Course) 15-780 Graduate Introduction to AI</a>
<ul>
<li>Recorded lectures <a href="https://www.youtube.com/playlist?list=PL_Ig1a5kxu55zgIn3Tb5Pc5C62Yo2sYHf">here</a> from prev term
</li>
<li>No formal pre-requisites except grad level standing so familiar with some undergrad calculus and linear algebra/probability
<ul>
<li>MIT's Artificial General Intelligence <a href="https://agi.mit.edu/">free class</a> explores possible research paths for creating AGI
</li>
<li>Sussman had a great 2017 grad <a href="https://ai6034.mit.edu/wiki/index.php?title=6.S966:_A_Graduate_Section_for_6.034#Prospectus">course</a> reviewing older paper with unsolved problems or deep ideas, like his own <i>Art of the Propagator</i> paper and <a href="http://groups.csail.mit.edu/mac/users/gjs/propagators/">implementation</a> </li>
<li>AI <a href="https://courses.csail.mit.edu/6.803/schedule.html">readings</a> to explain intelligence from a computational point of view
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div>
<h3>
<span>20.2</span> Math Background for ML</h3>
<p>
A series of <a href="https://www.youtube.com/playlist?list=PL7y-1rk2cCsAqRtWoZ95z-GMcecVG5mzA">recorded</a> lectures and recommended <a href="https://canvas.cmu.edu/courses/603/assignments/syllabus">texts</a> for a crash course in the Math background for grad level ML introductory courses
</p>
<div>
<h4>
<span>20.2.1</span> Statistics Theory</h4>
<div>
<ul>
<li>
<a href="http://www.stat.cmu.edu/~larry/=stat705/">(Full Course) 36-705 Intermediate Statistics</a>
<ul>
<li>The author of <i>All of Statistics</i> Larry Wasserman covers the fundamentals of theoretical statistics in these (badly) recorded lectures
</li>
<li>The probability prereqs can be found <a href="https://www.youtube.com/playlist?list=PL_Ig1a5kxu57qPZnHm-ie-D7vs9g7U-Cl">here</a> or from any undergrad text on probability
</li>
<li>Additional lecture <a href="http://www.stat.cmu.edu/~larry/=sml/">notes</a> on the intersection of stats theory and ML
</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div>
<h3>
<span>20.3</span> Introduction to ML</h3>
<div>
<ul>
<li>
<a href="http://www.cs.cmu.edu/~ninamf/courses/601sp15/lectures.shtml">(Full Course) 10-601 Masters Introduction to ML</a>
<ul>
<li>Recorded lectures and recitations
</li>
<li>Self contained, assumes you are grad level standing so have familiarity with basic probability and algebra
</li>
<li>Doesn't cover as much theory as 10-701/715
</li>
</ul>
</li>
</ul>
</div>
<div>
<h4>
<span>20.3.1</span> Practical Machine Learning</h4>
<div>
<p>
"This course will teach you basic skills to decide which learning algorithm to use for what problem, code up your own learning algorithm and evaluate and debug it."
This course compliments 10-601. Only prereq you may have not already done is Multivariate Taylor Series. You can directly look that up <a href="https://mml-book.github.io/">here</a> in the vector calculus chapter.
</p>
<ul>
<li>
<a href="http://www.cs.cornell.edu/courses/cs4780/2018fa/lectures/index.html">(Full Course) CS4780 Machine Learning for Intelligent Systems</a>
<ul>
<li>YouTube <a href="https://www.youtube.com/playlist?list=PLl8OlHZGYOQ7bkVbuRthEsaLr7bONzbXS">lectures</a>, additional reading <a href="http://www.cs.cornell.edu/courses/cs4780/2018fa/page18/index.html">recommendations</a> and videos for each topic
</li>
<li>Kaggle has numerous <a href="https://www.kaggle.com/learn/machine-learning">tutorials</a> and competitions available </li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div>
<h3>
<span>20.4</span> Advanced Introduction to ML</h3>
<div>
<ul>
<li>
<a href="http://www.cs.cmu.edu/~bapoczos/Classes/ML10715_2015Fall/index.html">(Full Course) 10-715 Advanced Introduction to ML</a> <ul>
<li>Most recent version <a href="https://sites.google.com/site/10715advancedmlintro2017f/lectures">here</a> uses mostly the same recorded lectures.
</li>
<li>Fast paced curriculum intended to prepare PhD students in the ML grad program to write research papers
</li>
<li>Some <a href="http://curtis.ml.cmu.edu/w/courses/index.php/Syllabus_for_Machine_Learning_with_Large_Datasets_10-605_in_Fall_2017">slides</a> on the practical techniques needed for working with large datasets
<ul>
<li>Linear Algebra review crash course w/recorded lectures <a href="http://www.cs.cmu.edu/~zkolter/course/linalg/index.html">here</a> relevant to optimization/ML
</li>
<li>Real Analysis review <a href="https://www.youtube.com/channel/UC7gOYDYEgXG1yIH_rc2LgOw/playlists">here</a>
</li>
<li>Probability lecture <a href="http://www.stat.cmu.edu/~siva/700/main.html">notes</a>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li>
<a href="http://www.cs.cmu.edu/~epxing/Class/10701/lecture.html">(Full Course) 10-701 Introduction to ML</a>
<ul>
<li>Intended for PhD students outside the ML program, more theory than 10-601 </li>
<li>Assumes you have this <a href="#sec-20-2">20.2</a> math background
</li>
<li>Some recorded recitations are <a href="https://www.youtube.com/playlist?list=PL_Ig1a5kxu54nrfuy4V0V1eHJk3C4dV3X">here</a>
<ul>
<li>Some <a href="http://curtis.ml.cmu.edu/w/courses/index.php/Syllabus_for_Machine_Learning_with_Large_Datasets_10-605_in_Fall_2017">slides</a> on the practical techniques needed for working with large datasets
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div>
<h3>
<span>20.6</span> Deep Learning</h3>
<div>
<p>
If you're interested in parallel GPU programming for training see these <a href="https://www.youtube.com/playlist?list=PLpIxOj-HnDsO4Atvrp86c-4La9Mq3kMQZ">lectures</a> and <a href="http://15418.courses.cs.cmu.edu/spring2016/lectures">notes</a>. </p>
<ul>
<li>
<a href="https://www.cs.cmu.edu/~bhiksha/courses/deeplearning/Fall.2015/">(Full Course) 11-785 Introduction to Deep Learning</a>
<ul>
<li>Recorded lectures <a href="https://www.youtube.com/playlist?list=PLwrNrH82f0KgdZhGB7sm9NrOanfWBS-EN">here</a> to learn to build and tune deep learning models
<ul>
<li>If the playlist is deleted, which is frequently, (use youtube-dl to archive) search YouTube for "CMU 11-785" </li>
</ul>
</li>
<li>There are also applied courses and practical challenges/competitions on kaggle.com
<ul>
<li>Some <a href="http://www.archive.ece.cmu.edu/~ece739/schedule.html">slides</a> on techniques for making deep learning robust to adversarial manipulation
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div>
<h3>
<span>20.8</span> Further Research</h3>
<div>
<ul>
<li>Kaggle <a href="https://www.kaggle.com/competitions">competitions</a>
</li>
<li>The <a href="http://www.jmlr.org/">Journal</a> of Machine Learning Research and other AI journals you can access with Sci-Hub
</li>
<li>Open <a href="http://www.chalearn.org/">challenges</a> in ML </li>
<li>
<a href="https://algorithmia.com/">Algorithmia.com</a> an open marketplace for AI algorithms
</li>
<li>A <a href="https://www.morganclaypool.com/toc/aim/1/1">series</a> of papers curated for AI you can read with sci-hub
</li>
</ul>
</div>
</div>
</div>
<div>
<h2>
<span>21</span> Graduate Research Elective: Cryptography</h2>
<p>
Read these <a href="http://www.cs.cmu.edu/~15751/2016-lecture1.pdf">slides</a> from <i>A Theorist's ToolKit</i> on how to find research, how to write math in LaTeX, how to give a talk, where to ask on stackexchange etc.
</p>
<div>
<h3>
<span>21.2</span> Applied Cryptography</h3>
<div>
<ul>
<li>Draft <a href="https://crypto.stanford.edu/~dabo/cryptobook/">book</a> <i>A Graduate Course in Applied Cryptography</i> - Dan Boneh and Victor Shoup
</li>
<li>Lecture <a href="https://www.ece.cmu.edu/~ece733/schedule.html">notes</a> from 18-733 Applied Cryptography </li>
<li>Read all Daniel J. Bernstein's (and Peter Gutmann's) posts on the IETF Crypto Forum Research Group [Cfrg] archive, it's a <a href="https://mailarchive.ietf.org/arch/msg/cfrg/E-q_6ABdKfPU5XG2N6IVAs_sNhA/?qid=435ba2be2c3db3444e338cc0259ee124">master class</a> in modern <a href="https://mailarchive.ietf.org/arch/msg/cfrg/oOY_QNA6QMsTJHLLbUjch6-GLsc/?qid=435ba2be2c3db3444e338cc0259ee124">cryptanalysis</a> and he rips apart bad standards/protocol designs. </li>
<li>Read <i>The Art of Computer Programming - Seminumerical Algorithms</i> by Knuth (Vol 2) chapter on Random Numbers. These tests are still used in MIT grad courses. Try them on every library you can find that supposedly generates pseudorandom numbers </li>
<li>
<a href="https://www.wireguard.com/formal-verification/">Read</a> about the proof of the Wireguard protocol, a VPN that uses <code>AEAD_CHACHA20_POLY1305</code>
</li>
</ul>
</div>
</div>
<div>
<h3>
<span>21.3</span> Future Research</h3>
<div>
<ul>
<li>Follow whatever the PhD students of djb and Tanja Lange are <a href="https://hyperelliptic.org/tanja/students.html">working on</a>
</li>
<li>
<a href="https://2017.pqcrypto.org/school/schedule.html">Watch</a> the lectures from the 2017 Post-Quantum Crypto Summer School </li>
<li>Read the journal of <a href="https://link.springer.com/journal/13389">Crypto Engineering</a> (use SciHub proxy)
</li>
<li>Read a <a href="http://www.math.cmu.edu/~af1p/BOOK.pdf">book</a> on Random Graphs there is a <a href="http://math.nsc.ru/conference/g2/g2c2/TokarevaN.pdf">connection</a> between Graph Theory and Cryptography </li>
<li>Try the <a href="https://cryptopals.com/">Cryptopals</a> challenges </li>
<li>Read some cryptocurrency papers, such as Stellar's Consensus algorithm (soon to be used by mobilecoin.com), <a href="https://medium.com/@homakov/introducing-failsafe-network-ea47ab476fe6">Fail-Safe Network</a> or the <a href="https://github.com/zcash/zips/blob/master/protocol/protocol.pdf">protocol specification</a> for Zcash.
</li>
</ul>
</div>
</div>
</div>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via functionalcs.github.io http://bit.ly/2IfR71u<br>
<br>
June 5, 2019 at 03:48PM
